linera-service/src/linera-exporter/storage.rs

```diff
@@ -0,0 +1,428 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{collections::BTreeMap, marker::PhantomData, sync::Arc};
+
+use dashmap::DashMap;
+use linera_base::{
+    crypto::CryptoHash,
+    data_types::{Blob, BlockHeight},
+    identifiers::{BlobId, ChainId},
+};
+use linera_chain::types::{ConfirmedBlock, ConfirmedBlockCertificate, GenericCertificate};
+use linera_client::config::{DestinationId, LimitsConfig};
+use linera_sdk::{
+    ensure,
+    views::{View, ViewError},
+};
+use linera_storage::Storage;
+use linera_views::{
+    batch::Batch, context::Context, log_view::LogView, store::WritableKeyValueStore as _,
+    views::ClonableView,
+};
+use mini_moka::unsync::Cache as LfuCache;
+use quick_cache::{sync::Cache as FifoCache, Weighter};
+
+use crate::{
+    common::{BlockId, ExporterError, LiteBlockId},
+    state::{BlockExporterStateView, DestinationStates},
+};
+
+pub(super) struct ExporterStorage<S>
+where
+    S: Storage + Clone + Send + Sync + 'static,
+{
+    shared_storage: SharedStorage<<S as Storage>::BlockExporterContext, S>,
+}
+
+struct SharedStorage<C, S>
+where
+    S: Storage + Clone + Send + Sync + 'static,
+{
+    storage: S,
+    destination_states: DestinationStates,
+    shared_canonical_state: CanonicalState<C>,
+    blobs_cache: Arc<FifoCache<BlobId, Arc<Blob>, BlobCacheWeighter>>,
+    blocks_cache: Arc<FifoCache<CryptoHash, Arc<ConfirmedBlockCertificate>, BlockCacheWeighter>>,
+}
+
+pub(super) struct BlockProcessorStorage<S>
+where
+    S: Storage + Clone + Send + Sync + 'static,
+{
+    shared_storage: SharedStorage<S::BlockExporterContext, S>,
+    chain_states_cache: LfuCache<ChainId, LiteBlockId>,
+    exporter_state_view: BlockExporterStateView<<S as Storage>::BlockExporterContext>,
+}
+
+impl<C, S> SharedStorage<C, S>
+where
+    C: Context + Send + Sync + 'static,
+    S: Storage + Clone + Send + Sync + 'static,
+{
+    fn new(
+        storage: S,
+        state_context: LogView<C, CryptoHash>,
```

Comment:
The problem here is that `state_context` does not indicate at all:
* That is a `View`,
* That is a `Log`.
* That it is storing hashes (of what?)